app/main.py

```diff
@@ -1,34 +1,121 @@
+        return decks
 
-    def fire(self, row, column):
-        # Change the `is_alive` status of the deck
-        # And update the `is_drowned` value if it's needed
-        pass
+    def get_deck(self,
+                 row: int,
+                 column: int
+                 ) -> Deck | None:
+        for deck in self.decks:
+            if deck.row == row and deck.column == column:
+                return deck
+        return None
+
+    def fire(self,
+             row: int,
+             column: int
+             ) -> None:
+        deck = self.get_deck(row, column)
+        if deck:
+            deck.is_alive = False
+            self.is_drowned = all(not d.is_alive for d in self.decks)
 
 
 class Battleship:
-    def __init__(self, ships):
-        # Create a dict `self.field`.
-        # Its keys are tuples - the coordinates of the non-empty cells,
-        # A value for each cell is a reference to the ship
-        # which is located in it
-        pass
-
-    def fire(self, location: tuple):
-        # This function should check whether the location
-        # is a key in the `self.field`
-        # If it is, then it should check if this cell is the last alive
-        # in the ship or not.
-        pass
+    def __init__(self,
+                 ships: list[tuple[tuple[int, int], tuple[int, int]]]
+                 ) -> None:
+        self.field = [["~" for _ in range(10)] for _ in range(10)]
+        self.ships = []
+        self.coord_to_ship = {}
+        for start, end in ships:
+            ship = Ship(start, end)
+            self.ships.append(ship)
+            for deck in ship.decks:
+                self.field[deck.row][deck.column] = "□"
+                self.coord_to_ship[(deck.row, deck.column)] = ship
+
+        self._validate_field()
+
+    def fire(self, ceil: tuple[int]) -> str:
+        if ceil not in self.coord_to_ship:
+            return "Miss!"
+        ship = self.coord_to_ship[ceil]
+        ship.fire(ceil[0], ceil[1])
+        if ship.is_drowned:
+            return "Sunk!"
+        return "Hit!"
+
+    def print_field(self) -> None:
+        for row in range(len(self.field)):
+            for column in range(len(self.field[0])):
+                ship = self.coord_to_ship.get((row, column))
+                if ship:
+                    deck = ship.get_deck(row, column)
+                    if deck.is_alive:
+                        print(u"\u25A1", end=" ")
+                    else:
+                        print("x", end=" ")
+                else:
+                    print("~", end=" ")
+            print()
+
+    def _validate_field(self) -> None:
+        ship_counts = {1: 0, 2: 0, 3: 0, 4: 0}
+        for ship in self.ships:
+            deck_count = len(ship.decks)
+            if deck_count > 4:
+                raise ValueError("Ship with more than 4 decks is not allowed.")
+            ship_counts[deck_count] += 1
+        if ship_counts != {1: 4, 2: 3, 3: 2, 4: 1}:
```

Comment:
The condition for ship distribution should match the specified requirements. Ensure that the ship distribution is {1: 4, 2: 3, 3: 2, 4: 1} as per the task requirements.